components/brave_ads/browser/ads_p2a.cc

```diff
@@ -16,37 +18,466 @@
 namespace brave_ads {
 namespace {
 
-const char kAdViewConfirmationCountHistogramName[] =
-    "Brave.P2A.ViewConfirmationCount";
-const uint16_t kAdViewConfirmationCountIntervals[] =
+const uint16_t kBinIntervals[] =
     { 0, 5, 10, 20, 50, 100, 250, 500 };
 
-void EmitAdViewConfirmationHistogram(uint64_t number_of_confirmations) {
-    const uint16_t* it = std::lower_bound(kAdViewConfirmationCountIntervals,
-    std::end(kAdViewConfirmationCountIntervals), number_of_confirmations);
-  const uint16_t answer = it - kAdViewConfirmationCountIntervals;
-  EmitConfirmationsCountMetric(answer);
-}
-
 }  // namespace
 
-void RegisterP2APrefs(PrefRegistrySimple* registry) {
-  registry->RegisterListPref(prefs::kAdViewConfirmationCountPrefName);
+void RegisterP2APrefs(
+    PrefRegistrySimple* registry) {
+  for (auto const& question_name : kP2AQuestionList) {
+      std::string pref_path(prefs::kP2AStoragePrefNamePrefix);
+      pref_path.append(question_name);
+      registry->RegisterListPref(pref_path);
+  }
 }
 
-void RecordEventInWeeklyStorage(
+void RecordInWeeklyStorageAndEmitP2AHistogramAnswer(
     PrefService* prefs,
-    const std::string& pref_name) {
-  if (pref_name == prefs::kAdViewConfirmationCountPrefName) {
-    WeeklyStorage storage(prefs, prefs::kAdViewConfirmationCountPrefName);
-    storage.AddDelta(1);
-    EmitAdViewConfirmationHistogram(storage.GetWeeklySum());
+    const std::string& name) {
+  std::string pref_path(prefs::kP2AStoragePrefNamePrefix);
+  pref_path.append(name);
+  if (!prefs->FindPreference(pref_path)) {
+    return;
+  }
+  WeeklyStorage storage(prefs, pref_path.c_str());
+  storage.AddDelta(1);
+  EmitP2AHistogramAnswer(name, storage.GetWeeklySum());
+}
+
+void EmitP2AHistogramAnswer(
+    const std::string& name,
+    int count_value) {
+  const uint16_t* it =
+      std::lower_bound(kBinIntervals, std::end(kBinIntervals), count_value);
+  const uint16_t answer = it - kBinIntervals;
+
+  if (name == "Brave.P2A.TotalAdOpportunities") {
+    UMA_HISTOGRAM_EXACT_LINEAR(
+        "Brave.P2A.TotalAdOpportunities",
+        answer, 8);
+  } else if (name == "Brave.P2A.AdOpportunitiesPerSegment.architecture") {
+    UMA_HISTOGRAM_EXACT_LINEAR(
```

Comment:
I reckon you can use functions from `histogram_functions.h` that allow non compile-time constant arguments